File: lib/util/subscription.dart
/// Service subscription. Call [cancel] to unsubscribe.
class Subscription {
  void Function()? _cancel;

  Subscription({required void Function() cancel}) : _cancel = cancel;

  /// Cancels the subscription.
  ///
  /// Can be called once. Subsequent calls have no effect.
  void cancel() {
    _cancel?.call();
    _cancel = null;
  }
}

/// Service subscription with async cancel. Call [cancel] to unsubscribe.
class AsyncSubscription {
  Future<void> Function()? _cancel;

  AsyncSubscription({required Future<void> Function() cancel})
      : _cancel = cancel;

  /// Cancels the subscription.
  ///
  /// Can be called once. Subsequent calls have no effect.
  Future<void> cancel() async {
    await _cancel?.call();
    _cancel = null;
  }
}

File: android/app/src/main/kotlin/mk/webfactory/flutter_template/platform_comm/Subscription.kt
package mk.webfactory.flutter_template.platform_comm;

/**
 * Service subscription. Call [cancel] to unsubscribe.
 */
class Subscription(private var cancel: (() -> Unit)?) {

    /**
     * Cancels the subscription.
     *
     * Can be called once. Subsequent calls have no effect.
     */
    fun cancel() {
        cancel?.invoke()
        cancel = null
    }
}